### PR TITLE
Hd/fix camera and light reset and copy

### PR DESCRIPTION
**Pull request by RSlysz**
### Purpose of this PR
CopyTo method in Camera and light have not been maintened from long time. This PR will add missing copy. (Fix https://fogbugz.unity3d.com/f/cases/1330194/)
Also it introduce a test on some classes that have a CopyTo method.

The test in itself is a little rough as we must create the object and a mutated version of the object, try to use CopyTo one onto the other and then check that each fields are equals.
So to be sure this huge logic of a test work as intended, there is self test in it.
image
A: self tests (checking inherited field, filtered fields, and most of the type)
B: class that have a CopyTo and that are tested

---
### Testing status
Automated added self testing are green.
Automated added testing on each class that have a CopyTo are green.
Used the test to find missing copy in CopyTo and also needed white listing. (so there should be no false positive)

---
### Comments to reviewers
Check [CopyFilterAttribute](https://github.com/Unity-Technologies/Graphics/pull/4415/files#diff-d4070c590a78fd1d0b2b8e7cdbde18f0033cd8838570876665e554e0352f145e) comment to have usage of the filtering mechanism for white listing
@pmavridis if you see anything that should not be here for the copy of Camera, tell us :)
@alelievre if you see anything that should not be here for the copy of Light, tell us :)
@fredericv-unity3d , @alex-vazquez if you see anyway to make this less painful to maintain, or if you see missing cases or wrong use of some reflection method... I need experts to see better here.

### CODE REVIEW

**Review by fredericv-unity3d — APPROVED** [review]
Approved, this automatic test is required for such method. Good job on adding it.

Regarding maintenance, I only see one way to remove the burden and human mistakes from the maintenance of this method, it is to parse the C# files and generate the appropriate code.

**Comment by fredericv-unity3d** [comment]
Regarding code generation, if you are interested I have one in my [current work](https://github.cds.internal.unity3d.com/fredericv/GameObjectPooling/tree/master/Tools).

It parses C# files to find classes and specific attributes and generate custom classes, serialized versions, editor, property drawers and other few things. It is triggered by a running a bat file (runs a dotnet build & run)

**Review by pmavridis — APPROVED** [review]

**Comment by RSlysz** [comment]
> 
> 
> Regarding code generation, if you are interested I have one in my [current work](https://github.cds.internal.unity3d.com/fredericv/GameObjectPooling/tree/master/Tools).
> 
> It parses C# files to find classes and specific attributes and generate custom classes, serialized versions, editor, property drawers and other few things. It is triggered by a running a bat file (runs a dotnet build & run)

I fear that the dev adding a new field will not take fully this into account. If the field is "deep copied" object which should have not same reference but same content, he can forget to specify it. Same for parameters for debug or that should be reconstructed for the instance and not be copied.

I take not of it and will try to fix it this way. Though, it is too short to do it before the incoming release and fully test it regarding above. (At least, I have not the time to do this.)

**Review by remi-chapelain — CHANGES_REQUESTED** [review]
Quickly checked every properties of the camera:
The only thing that does not properly reset it's the unit of the shutter speed physical camera. The default unit is in "second" but if you change it to "1/second" and then reset the camera component, the value is properly reset but the unit stays. This is **not a big issue** because even though the unit stays, the value itself is still reset at its default. 
Won't block the PR for this. 

https://user-images.githubusercontent.com/57442369/117003269-edf96e80-ace4-11eb-932d-8d3e29b39b5b.mp4

Did also the same for lights and there's two little things (that also repro on latest packages, so not introduced by this PR)
- When resetting the light component, intensity of the light is set to 0, not sure that's what we want (default intensity for point light when creating them is 600)
- When resetting, the cookie texture stays. 

d229d918eee57962be103553195f1614

**Comment by RSlysz** [comment]
>  the shutter speed physical camera

It is actually not something in the component Camera of HDAdditionalCameraData. It is something that is saved in EditorPref. As such, it is shared amongst all Camera. I dunno why it is done this way. If need to change, please create a FB on this as it need to investigate why it is done this way.

>     * When resetting the light component, intensity of the light is set to 0, not sure that's what we want (default intensity for point light when creating them is 600)
> 
>     * When resetting, the cookie texture stays.

I'm checking thoses two as they indeed seems odd.

**Review by alelievr — APPROVED** [review]

**Comment by RSlysz** [comment]
Above is the missing intensity init in the reset of the light. CopyTo is unchanged.

Missing cookie reset is a miss in the built-in Camera.

Also on type of master, there is a strange error happening after reseting.
`Awake has not been called 'Spot Light' (Light). Figure out where the object gets created and call AwakeFromLoad correctly.`
This is send by c++ and should only be send when the component is used without doing any Awake. As it was not there on a13 where I tested initially, it should be a regression on lasts trunk.

**Comment by RSlysz** [comment]
Above is a patch for reseting built-in light cookie on our side.

**Review by remi-chapelain — APPROVED** [review]
Issue with light intensity and cookie not being cleared fixed ! :) 

🟢

**Comment by RSlysz** [comment]
@sebastienlagarde Alex is right. I can refactor a little the code to be more simple to use in the future. **Don't merge right now.** I'll update it as soon as I can.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs
Comment (RSlysz) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:545]: Weird, I can't find that one. And the test can't find it too. It is a new field that will come if I merge ?

**Comment by RSlysz** [comment]
@alex-vazquez I integrated the small attributes, it is more convenient. However the interface is just causing more issue than it solve due to the type change in the CopyTo argument.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs
Comment (alex-vazquez) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDAdditionalCameraData.cs:576]: I did a rebase while I was trying your PR, maybe?

**Review by JulienIgnace-Unity — APPROVED** [review]

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**
